### PR TITLE
[Rename] Display rename information on package detail page

### PR DESCRIPTION
**Pull request by zhhyu**
image
image

1. Eager loading the "ToPackageRegistration" info.
2. Reuse CSS template from deprecations;
3. Feature Flag.

### CODE REVIEW

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 1]
```diff
@@ -1486,11 +1622,17 @@ public async Task UsesProperIconUrl()
...
+                    .Setup(x => x.GetPackageRenames(package.PackageRegistration))
+                    .Verifiable();
+
                 var result = await controller.DisplayPackage(id, version: null);
                 var model = ResultAssert.IsView<DisplayPackageViewModel>(result);
                 iconUrlProvider
                     .Verify(iup => iup.GetIconUrlString(package), Times.AtLeastOnce);
                 Assert.Equal(iconUrl, model.IconUrl);
+                renameService.Verify();
+
```
Comment (loic-sharma) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs:1635]: There's an empty new line here

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 2]
```diff
@@ -1378,6 +1445,66 @@ public async Task ShowsFirstDeprecationPerPackage()
...
+                renameService
+                    .Setup(x => x.GetPackageRenames(package.PackageRegistration))
+                    .Verifiable();
+
+                // Act
+                var result = await controller.DisplayPackage(id, version: null);
+
+                // Assert
+                var model = ResultAssert.IsView<DisplayPackageViewModel>(result);
+                Assert.Equal(isRenamesEnabledForThisUser, model.IsPackageRenamesEnabled);
```
Comment (loic-sharma) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs:1504]: Could we assert that renames returned by the rename service are added to the view model?

Comment (zhhyu) [reply tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs:1508]: Good point! Add it.

File: src/NuGetGallery/Scripts/gallery/page-display-package.js [hunk 3]
```diff
@@ -1,6 +1,15 @@
 $(function () {
     'use strict';
 
+    // Configure the rename information container
+    var container = $('#show-rename-content-container');
+    window.nuget.configureExpander("rename-content-container", "ChevronDown", null, "ChevronUp");
+    container.keydown(function (event) {
```
Comment (joelverhagen) [inline src/NuGetGallery/Scripts/gallery/page-display-package.js:7]: This should be DRY with the deprecation below. Shared function, loop, whatever.

File: src/NuGetGallery/Views/Packages/_DisplayPackageRenames.cshtml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        <div id="show-rename-content-container" class="deprecation-expander" tabindex="0" data-toggle="collapse" data-target="#rename-content-container" aria-expanded="false" aria-controls="rename-content-container" aria-labelledby="rename-container-label" role="button">
+            <div class="deprecation-expander" role="button">
+                <div class="deprecation-expander-container">
+                    <i class="deprecation-expander-icon ms-Icon ms-Icon--Info" aria-hidden="true"></i>
+
+                    <div id="rename-container-label" class="deprecation-expander-info-right">
+                        @{
+                            if (Model.PackageRenames.Count == 1)
+                            {
+                                @:This package has been renamed
```
Comment (joelverhagen) [inline src/NuGetGallery/Views/Packages/_DisplayPackageRenames.cshtml:14]: Period at the end of sentence.

File: src/NuGetGallery/Views/Packages/_DisplayPackageRenames.cshtml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+                            }
+                            else
+                            {
+                                @:This package has been renamed or split into new packages
```
Comment (joelverhagen) [inline src/NuGetGallery/Views/Packages/_DisplayPackageRenames.cshtml:18]: Period at the end of sentence.

File: src/NuGetGallery/Services/PackageRenameService.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        private readonly IEntityRepository<PackageRename> _packageRenameRepository;
+
+        public PackageRenameService (
+            IEntityRepository<PackageRename> packageRenameRepository)
+        {
+            _packageRenameRepository = packageRenameRepository ?? throw new ArgumentNullException(nameof(packageRenameRepository));
+        }
+
+        public IReadOnlyList<PackageRename> GetPackageRenames(PackageRegistration packageRegistration)
+        {
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/PackageRenameService.cs:23]: Will this new query significantly impact the display package page performance?

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 7]
```diff
@@ -837,11 +840,14 @@ public virtual async Task<ActionResult> DisplayPackage(string id, string version
                 .GroupBy(d => d.PackageKey)
                 .ToDictionary(g => g.Key, g => g.First());
 
+            var packageRenames = _renameService.GetPackageRenames(package.PackageRegistration);
```
Comment (joelverhagen) [inline src/NuGetGallery/Controllers/PackagesController.cs:843]: This should be behind the flag as well, so we can turn if off if perf is an issue.

**Comment by zhhyu** [comment]
It's behind the feature flag now. If the feature flag is turned off, there will be no impact on loading the package page, and nothing related to rename will be triggered.
I prefer to check in this change, but until we have the results from load test or A/B test, we should not turn it on for customers.

**Pull request merged**